src/Mod/Sketcher/Gui/ViewProviderSketch.cpp

```diff
@@ -3481,6 +3526,224 @@ QIcon ViewProviderSketch::mergeColorfulOverlayIcons (const QIcon & orig) const
     return Gui::ViewProvider::mergeColorfulOverlayIcons (mergedicon);
 }
 
+bool ViewProviderSketch::getCameraMaxDimension(const Gui::View3DInventorViewer* viewer) {
+    float fHeight = -1.0;
+    float fWidth = -1.0;
+    viewer->getDimensions(fHeight, fWidth);
+    if (fHeight < 0 || fWidth < 0)
+        return false;
+
+    fHeight = std::max(fHeight, fWidth);
+    if (fabs(fHeight - camMaxDimension) > 0) { //ie if user zoomed.
+        camMaxDimension = fHeight;
+        return true;
+    }
+    return false;
+}
+
+bool ViewProviderSketch::getCameraCenter(const Gui::View3DInventorViewer* viewer)
+{
+    SoCamera* cam = viewer->getSoRenderManager()->getCamera();
+    if(!cam)
+        return false;
+
+    SbVec3f direction;
+    cam->orientation.getValue().multVec(SbVec3f(0, 0, -1), direction);
+    SbVec3f newCamCenterOnSketch = cam->position.getValue() + cam->focalDistance.getValue() * direction;
+
+    if ((camCenterOnSketch - newCamCenterOnSketch).length() > 0.1 * camMaxDimension) { //ie if user moved by more than 10% of the current 3d view dimensions.
+        camCenterOnSketch = newCamCenterOnSketch;
+        return true;
+    }
+    return false;
+}
+
+double ViewProviderSketch::getGridSize(const Gui::View3DInventorViewer* viewer)
+{
+    short pixelWidth = -1;
+    short pixelHeight = -1;
+    viewer->getViewportRegion().getViewportSizePixels().getValue(pixelWidth, pixelHeight);
+    if (pixelWidth < 0 || pixelHeight < 0)
+        return GridSize.getValue();
+
+    ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Mod/Sketcher/General");
+    int numberOfLines = std::max(pixelWidth, pixelHeight) / hGrp->GetInt("GridSizePixelThreshold", 15);
+    ParameterGrp::handle hGrp2 = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Units");
+    int units = hGrp2->GetInt("UserSchema", 0); //2 3 5 7 are imperial schemas. 2 3 inches, 5 7 feets
+    double unitMultiplier = (units == 2 || units == 3) ? 25.4 : (units == 5 || units == 7) ? 304.8 : 1;
+
+    double newGridSize = unitMultiplier * pow(hGrp->GetInt("GridNumberSubdivision", 10), 1 + floor(log10(camMaxDimension / unitMultiplier / numberOfLines)));

```

Comment:
Testing with GridNumberSubdivision=5 I realized that the grid gets very tight before it resizes, this is because we are using log10 here which works excellently when there is 10 subdivisions, but it should be the logarithm with base=GridNumberSubdivision in the general case.
This is embarassing but I can't remember now the math to get a different base logarithm using natural or base 10 logarithms 😅

Reply:
this seems to work
```suggestion
    int subdivisions = hGrp->GetInt("GridNumberSubdivision", 10);
    double newGridSize = unitMultiplier * pow(subdivisions, 1 + floor(log(camMaxDimension / unitMultiplier / numberOfLines) / log(subdivisions)));

```
keeping in mind I'm a C++ noob, do we need to declare `subdivisions` as `const int` or is this irrelevant?

Reply:
I will test that to check if it works as advertized.
I don't know about const, not sure when it should be used but I don't think it's needed here. That said I'm also mostly a C++ noob :)

Reply:
Works, this fix the threshold unstable when subdivision is not 10.